doc/src/build/programming-with-objects/ch5-dynamic-fields.md

```diff
@@ -0,0 +1,166 @@
+---
+title: Chapter 5 - Dynamic Fields
+---
+
+In previous chapters, we walked through various ways to use object fields to store primitive data and other objects (wrapping), but there are a few limitations to this approach:
+
+1. Object's have a finite set of fields keyed by identifiers that are fixed when its module is published.
+2. An object can become very large if it wraps several other objects. Larger objects can lead to higher gas fees in transactions. In addition, there is an upper bound on object size.
+3. As we will see in future chapters, there will be use cases where we need to store a collection of objects of heterogeneous types. Since the Move `vector` type must be templated on one single type `T`, it is not suitable for this.
+
+Fortunately, Sui provides *dynamic fields* with arbitrary names (not just identifiers), added and removed on-the-fly (not fixed at publish), which only affect gas when they are accessed, and can store heterogeneous values. This chapter introduces the libraries for interacting with this kind of field.
+
+### Fields vs Object Fields
+
+There are two flavors of dynamic field -- "fields" and "object fields" -- which differ based on how their values are stored:
+
+- **Object field** values *must* be objects themselves (have the `key` ability, and `id: UID` as the first field) and will still appear under their ID to external tools (explorers, wallets, etc) accessing storage even when stored in a dynamic object field.
+- **Field** values do not need to be objects. An object stored in a (non-object) field will not be accessible via its ID while it is stored in a dynamic field.
+
+The interfaces for interacting with these fields can be found in the [`dynamic_object_field`](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/sources/dynamic_object_field.move) and [`dynamic_field`](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/sources/dynamic_field.move) libraries respectively.
+
+### Field Names
+
+Unlike an object's regular fields whose names must be Move identifiers, dynamic field names can be any value that has `copy`, `drop` and `store`. This includes all Move primitives (integers, booleans, byte strings), and structs whose contents all have `copy`, `drop` and `store`.
+
+### Adding Dynamic Fields
+
+Dynamic fields are added with the following APIs:
+
+```rust
+use sui::dynamic_field as field;
+use sui::dynamic_object_field as ofield;
+
+public fun field::add<Name: copy + drop + store, Value: store>(
```

Comment:
I get what's going on here... but still a bit strange as it isn't valid Move syntax. And might be confusing for someone without a programming language background where such a syntax makes sense. 

Maybe
```
module sui::dynamic_field {
...
public fun add<Name: copy + drop + store, Value: store>(
  object: &mut UID,
  name: Name,
  value: Value,
)
}
```
And similarly for `dynamic_object_field`